### PR TITLE
initial implementation of Map::setView

### PR DESCRIPTION
**Pull request by mikemorris**
I'm not sure if something needs to change in `MapContext` or if I'm just implementing wrong, but calling `context->invokeSync(&MapContext::setView, *view);` results in the error `field type 'mbgl::View' is an abstract class`. Would love a quick review here from @kkaefer or @tmpsantos to check if I'm going in the right direction.

### CODE REVIEW

**Comment by mikemorris** [comment]
Got past the `invoke` trouble thanks to @tmpsantos' suggestion to make everything `View*`

File: src/mbgl/map/map_context.cpp [hunk 1]
```diff
@@ -50,8 +47,15 @@ MapContext::~MapContext() {
     assert(!style);
 }
 
+void MapContext::setView(View* view_) {
+    assert(util::ThreadContext::currentlyOn(util::ThreadType::Map));
+
+    view = view_;
```
Comment (tmpsantos) [inline src/mbgl/map/map_context.cpp]: You must call here `view->deactivate()` before replacing.

File: src/mbgl/map/map.cpp [hunk 2]
```diff
@@ -43,22 +43,31 @@ void Map::resume() {
     paused = false;
 }
 
+void Map::setView(View* view_) {
+    view = view_;
+    transform = std::make_unique<Transform>(*view);
+    context->invoke(&MapContext::setView, view);
```
Comment (tmpsantos) [inline src/mbgl/map/map.cpp]: You might need to make this a `context->invokeSync` to make sure it is safe to reuse the old view (i.e. `deactivate()` was called) in another map object right after calling this method.

**Comment by jfirebaugh** [comment]
In https://github.com/mapbox/node-mapbox-gl-native/issues/142 I'm advocating we abandon this and take a different approach.

**Comment by mikemorris** [comment]
-> https://github.com/mapbox/mapbox-gl-native/pull/2310

**Pull request closed without merging**